File: clients/python/hydracept/jobs.py
"""JobRunner — submit, nextAction, poll, download, receipt for capability jobs."""

from __future__ import annotations

import hashlib
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, TYPE_CHECKING

if TYPE_CHECKING:
    from hydracept import HydraceptClient
    from hydracept.cli.workspace import ResolvedWorkspace

_MEDIA_EXTENSIONS = {
    "audio/ogg": ".ogg",
    "application/ogg": ".ogg",
    "audio/wav": ".wav",
    "audio/mpeg": ".mp3",
    "image/png": ".png",
    "image/jpeg": ".jpg",
    "image/webp": ".webp",
    "model/gltf-binary": ".glb",
    "video/mp4": ".mp4",
    "application/json": ".json",
}


def normalize_sha256_digest(value: str | None) -> str:
    """Return lowercase hex, stripping a ``sha256:`` prefix when present."""
    raw = str(value or "").strip().lower()
    if raw.startswith("sha256:"):
        raw = raw[7:]
    return raw


def filename_for_download(item: dict[str, Any], artifact_id: str) -> str:
    media = str(item.get("mediaType") or item.get("mimeType") or "").split(";")[0].strip().lower()
    raw = str(item.get("filename") or item.get("label") or artifact_id or "artifact").strip() or "artifact"
    name = Path(raw.replace("\\", "/")).name or "artifact"
    ext = _MEDIA_EXTENSIONS.get(media, "")
    if ext and not name.lower().endswith(ext):
        name = f"{name}{ext}"
    return name


def next_action_for_job(job: dict[str, Any], *, has_artifacts: bool = False) -> str:
    status = str(job.get("status") or "").lower()
    if status in {"queued", "running", "awaiting_approval", "canceling"}:
        return "poll"
    if status == "succeeded":
        if has_artifacts:
            return "download"
        return "receipt"
    return "done"


@dataclass
class JobRunResult:
    job_id: str
    status: str
    job: dict[str, Any]
    receipt: dict[str, Any] | None
    downloads: list[Path] = field(default_factory=list)
    next_action: str = "poll"


class JobRunner:
    """Capability-job orchestration. Does not speak invocations or visual jobs."""

    def __init__(self, client: HydraceptClient, workspace: ResolvedWorkspace) -> None:
        self._client = client
        self._workspace = workspace

    def submit(self, capability_key: str, body: dict[str, Any]) -> dict[str, Any]:
        from hydracept.cli.job_context import merge_workspace_job_context

        payload = merge_workspace_job_context(dict(body), self._workspace)
        return self._client.submit_capability_job(capability_key, payload)

    def poll(self, job_id: str) -> dict[str, Any]:
        return self._client.get_job(job_id)

    def receipt(self, job_id: str) -> dict[str, Any]:
        return self._client.get_job_receipt(job_id)

    def next_action(self, job: dict[str, Any]) -> str:
        artifacts = job.get("artifacts") or []
        has_artifacts = isinstance(artifacts, list) and bool(artifacts)
        return next_action_for_job(job, has_artifacts=has_artifacts)

    def run(
        self,
        capability_key: str,
        body: dict[str, Any],
        *,
        download_dir: str | Path | None = None,
        poll_seconds: float = 120.0,
        poll_interval: float = 2.0,
    ) -> JobRunResult:
        submitted = self.submit(capability_key, body)
        job_id = str(submitted.get("jobId") or submitted.get("id") or "")
        if not job_id:
            raise RuntimeError(f"No jobId in submit response: {submitted!r}")
        deadline = time.time() + max(5.0, poll_seconds)
        job = submitted
        status = str(job.get("status") or "queued")
        while time.time() < deadline:
            job = self.poll(job_id)
            status = str(job.get("status") or "unknown")
            if status in {"succeeded", "failed", "canceled", "cancelled"}:
                break
            time.sleep(max(0.2, poll_interval))
        receipt: dict[str, Any] | None = None
        if status == "succeeded":
            try:
                receipt = self.receipt(job_id)
            except Exception:  # noqa: BLE001
                receipt = None
        downloads: list[Path] = []
        if download_dir and status == "succeeded":
            downloads = self.download_artifacts(job_id, receipt or job, Path(download_dir))
        has_artifacts = bool(downloads) or bool((receipt or {}).get("artifacts"))
        return JobRunResult(
            job_id=job_id,
            status=status,
            job=job,
            receipt=receipt,
            downloads=downloads,
            next_action=next_action_for_job(job, has_artifacts=has_artifacts),
        )

    def download_artifacts(
        self,
        job_id: str,
        receipt_or_job: dict[str, Any],
        download_dir: Path,
    ) -> list[Path]:
        download_dir.mkdir(parents=True, exist_ok=True)
        artifacts = receipt_or_job.get("artifacts") or []
        paths: list[Path] = []
        if not isinstance(artifacts, list):
            return paths
        for item in artifacts:
            artifact_id = ""
            filename = ""
            expected_sha = ""
            if isinstance(item, dict):
                artifact_id = str(item.get("artifactId") or item.get("id") or "")
                filename = filename_for_download(item, artifact_id)
                nested = item.get("digest") if isinstance(item.get("digest"), dict) else {}
                expected_sha = normalize_sha256_digest(
                    str(item.get("sha256") or nested.get("sha256") or "")
                )
            elif item:
                artifact_id = str(item)
                filename = artifact_id
            if not artifact_id:
                continue
            data = self._client.download_job_artifact(job_id, artifact_id)
            actual_sha = hashlib.sha256(data).hexdigest()
            if expected_sha and expected_sha != actual_sha:
                raise RuntimeError(
                    f"SHA-256 mismatch for {artifact_id}: receipt={expected_sha} downloaded={actual_sha}"
                )
            target = download_dir / Path(filename).name
            target.write_bytes(data)
            paths.append(target)
        return paths

File: clients/typescript/src/jobs.ts
import { createHash } from 'node:crypto';
import { mkdirSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import type { HydraceptClient } from './index.js';

export type HydraceptJobBinding = {
  projectId: string;
  environment?: string;
};

export type HydraceptJobRunResult = {
  jobId: string;
  status: string;
  job: Record<string, unknown>;
  receipt: Record<string, unknown> | null;
  downloads: string[];
  nextAction: string;
};

function nextActionForJob(job: Record<string, unknown>): string {
  const status = String(job.status || '').toLowerCase();
  const artifacts = job.artifacts;
  const hasArtifacts = Array.isArray(artifacts) && artifacts.length > 0;
  if (['queued', 'running', 'awaiting_approval', 'canceling'].includes(status)) return 'poll';
  if (status === 'succeeded') return hasArtifacts ? 'download' : 'receipt';
  return 'done';
}

export type HydraceptJobRunner = {
  run(
    capabilityKey: string,
    body: Record<string, unknown>,
    options?: { downloadDir?: string; outputRoot?: string; pollMs?: number },
  ): Promise<HydraceptJobRunResult>;
};

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

const MEDIA_EXTENSIONS: Record<string, string> = {
  'audio/ogg': '.ogg',
  'application/ogg': '.ogg',
  'audio/wav': '.wav',
  'audio/mpeg': '.mp3',
  'image/png': '.png',
  'image/jpeg': '.jpg',
  'image/webp': '.webp',
  'model/gltf-binary': '.glb',
  'video/mp4': '.mp4',
  'application/json': '.json',
};

export function normalizeSha256Digest(value: string | undefined | null): string {
  let raw = String(value || '').trim().toLowerCase();
  if (raw.startsWith('sha256:')) raw = raw.slice(7);
  return raw;
}

function filenameForDownload(item: Record<string, unknown>, artifactId: string): string {
  const media = String(item.mediaType || item.mimeType || '')
    .split(';')[0]
    .trim()
    .toLowerCase();
  let name = String(item.filename || item.label || artifactId || 'artifact')
    .replace(/\\/g, '/')
    .split('/')
    .pop() || 'artifact';
  const ext = MEDIA_EXTENSIONS[media];
  if (ext && !name.toLowerCase().endsWith(ext)) name = `${name}${ext}`;
  return name;
}

function omitStaleQuoteIds(body: Record<string, unknown>): Record<string, unknown> {
  const execution = body.execution;
  if (!execution || typeof execution !== 'object') return body;
  const cleaned = { ...(execution as Record<string, unknown>) };
  delete cleaned.quoteId;
  delete cleaned.estimateId;
  return { ...body, execution: cleaned };
}

export function createHydraceptJobRunner(
  client: HydraceptClient,
  binding: HydraceptJobBinding,
): HydraceptJobRunner {
  return {
    async run(capabilityKey, body, options = {}) {
      const payload = omitStaleQuoteIds({
        ...body,
        context: {
          productId: binding.projectId,
          projectId: binding.projectId,
          environment: binding.environment || 'development',
          ...((body.context as Record<string, unknown> | undefined) ?? {}),
        },
      });
      const submitted = (await client.runtime.submitCapabilityJob(capabilityKey, payload)) as {
        jobId?: string;
        id?: string;
        status?: string;
      };
      const jobId = String(submitted.jobId || submitted.id || '');
      if (!jobId) throw new Error('No jobId in submit response');
      const dest =
        options.downloadDir ||
        (options.outputRoot ? join(options.outputRoot, '.hydracept', 'output', jobId) : undefined);
      const pollMs = options.pollMs ?? 120_000;
      const deadline = Number.isFinite(pollMs) ? Date.now() + pollMs : Number.POSITIVE_INFINITY;
      let job: Record<string, unknown> = submitted as Record<string, unknown>;
      let status = String(submitted.status || 'queued');
      while (true) {
        job = (await client.runtime.getJob(jobId)) as Record<string, unknown>;
        status = String(job.status || 'unknown');
        if (['succeeded', 'failed', 'canceled', 'cancelled'].includes(status)) break;
        if (Date.now() >= deadline) break;
        await sleep(2000);
      }
      let receipt: Record<string, unknown> | null = null;
      if (status === 'succeeded') {
        receipt = (await client.runtime.getJobReceipt(jobId)) as Record<string, unknown>;
      }
      const downloads: string[] = [];
      if (dest && status === 'succeeded' && receipt) {
        const artifacts = (receipt.artifacts as Array<Record<string, unknown>>) || [];
        for (const item of artifacts) {
          const artifactId = String(item.artifactId || item.id || '');
          if (!artifactId) continue;
          const buffer = await client.http.fetchBytes(
            `/jobs/${encodeURIComponent(jobId)}/artifacts/${encodeURIComponent(artifactId)}`,
          );
          const bytes = new Uint8Array(buffer);
          const nested =
            item.digest && typeof item.digest === 'object'
              ? (item.digest as Record<string, unknown>)
              : {};
          const expected = normalizeSha256Digest(
            String(item.sha256 || nested.sha256 || ''),
          );
          const actual = createHash('sha256').update(bytes).digest('hex');
          if (expected && expected !== actual) {
            throw new Error(`SHA-256 mismatch for ${artifactId}`);
          }
          const filename = filenameForDownload(item, artifactId);
          const target = join(dest, filename);
          mkdirSync(dirname(target), { recursive: true });
          writeFileSync(target, bytes);
          downloads.push(target);
        }
      }
      return { jobId, status, job, receipt, downloads, nextAction: nextActionForJob(job) };
    },
  };
}
